client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,417 @@
+    static mbedtls_aes_context aes_ctx;
+
+    if (!initialized) {
+        mbedtls_aes_init(&aes_ctx);
+        initialized = true;
+    }
+
+    if (mode == MBEDTLS_AES_ENCRYPT) {
+        mbedtls_aes_setkey_enc(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_ENCRYPT, input, output);
+    } else if (mode == MBEDTLS_AES_DECRYPT) {
+        mbedtls_aes_setkey_dec(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_DECRYPT, input, output);
+    }
+
+    return KAA_ERR_NONE;
+}
+
+static int init_aes_key(unsigned char *key, size_t bytes)
+{
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_entropy_context entropy;
+
+    const uint8_t pers[] = "aes_generate_key";
+    int ret;
+
+    mbedtls_entropy_init(&entropy);
+    mbedtls_ctr_drbg_init(&ctr_drbg);
+
+    if ((ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy,
+            pers, sizeof(pers) - 1)) == 0) {
+        ret = mbedtls_ctr_drbg_random(&ctr_drbg, key, bytes);
+    }
+
+    mbedtls_ctr_drbg_free(&ctr_drbg);
+    mbedtls_entropy_free(&entropy);
+
+    return ret;
+}
+
+/* Performs initialization of the keys */
+kaa_error_t kaa_init_keys(void)
+{
+    /* Initialization should be performed only once */
+    static bool initialized = false;
+    if (!initialized) {
+        /* Generate session key */
+        if (init_aes_key(keys.session_key, KAA_SESSION_KEY_LENGTH)) {
+            return KAA_ERR_BADDATA;
+        }
+        keys.session_key_length = KAA_SESSION_KEY_LENGTH;
+        if (mbedtls_pk_parse_keyfile(&pk_context_, KAA_PRIVATE_KEY_STORAGE, NULL)) {
+#ifdef KAA_RUNTIME_KEY_GENERATION
+            if (rsa_genkey(&pk_context_)) {
+                return -1;
+            }
+#else
+            return KAA_ERR_BADDATA;
+#endif
+        }
+         if (mbedtls_pk_parse_public_keyfile(&pk_pub_context, KAA_PUBLIC_KEY_STORAGE)) {
+#ifdef KAA_RUNTIME_KEY_GENERATION
+		pk_pub_context = pk_context_;
+#else
+            return KAA_ERR_BADDATA;
+#endif
+        }
+        initialized = true;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+void  kaa_deinit_keys(void)
+{
+    mbedtls_pk_free(&pk_context_);
+}
+
+void ext_get_endpoint_session_key(uint8_t **buffer, size_t *buffer_size)
+{
+    if (!buffer || !buffer_size) {
+        return;
+    }
+    *buffer = keys.session_key;
+    *buffer_size = keys.session_key_length;
+}
+
+kaa_error_t ext_encrypt_data(const uint8_t *input, size_t payload_size, uint8_t *output)
+{
+    if (!input || !payload_size || !output) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    /* Adding PKCS7 padding */
+    size_t enc_data_size = ext_get_encrypted_data_size(payload_size);
+    uint8_t padding = enc_data_size - payload_size;
+    memset(output + payload_size, padding, padding);
+
+    /* Process the buffer chunk by chunk */
+    while (enc_data_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_ENCRYPT, input,
```

Comment:
You can check this against `KAA_ERR_NONE` to emphasize that it is a Kaa SDK routine